SFS_ir/interpolate_ir.m

```diff
@@ -88,71 +91,41 @@
-            % Magnitude and phase will be interpolated separately
-            magnitude = abs(TF);
-            phase = unwrap(angle(TF),[],3);
-            % Calculate interpolation only for the first half of the spectrum
-            % and only for original bins
-            idx_half = floor(size(TF,3)/2)+1;
-            magnitude_new(1,:) = interpolation(...
-                squeeze(magnitude(1:2,1,1:4:idx_half))',x0(:,1:2),xs);
-            magnitude_new(2,:) = interpolation(...
-                squeeze(magnitude(1:2,2,1:4:idx_half))',x0(:,1:2),xs);
-            phase_new(1,:) = interpolation(...
-                squeeze(phase(1:2,1,1:4:idx_half))',x0(:,1:2),xs);
-            phase_new(2,:) = interpolation(...
-                squeeze(phase(1:2,2,1:4:idx_half))',x0(:,1:2),xs);
-            % Calculate interpolated impulse response from magnitude and phase
-            ir_new(1,1,:) = ifft(magnitude_new(1,:) ...
-                .* exp(1i*phase_new(1,:)),size(ir,3),'symmetric');
-            ir_new(1,2,:) = ifft(magnitude_new(2,:)...
-                .* exp(1i*phase_new(2,:)),size(ir,3),'symmetric');
-        otherwise
-            error('%s: %s is an unknown interpolation method.', ...
-                upper(mfilename),interpolationmethod);
-        end
-    else
-        % --- 2D interpolation ---
-        warning('SFS:irs_intpol3D',...
-            ['doing 2D IR interpolation between (%.1f,%.1f) deg, ', ...
-             '(%.1f,%.1f) deg and (%.1f,%.1f) deg.'], ...
-            deg(x0(1,1)), deg(x0(2,1)), ...
-            deg(x0(1,2)), deg(x0(2,2)), ...
-            deg(x0(1,3)), deg(x0(2,3)));
-        ir_new(1,1,:) = interpolation(squeeze(ir(1:3,1,:))',x0(:,1:3),xs);
-        ir_new(1,2,:) = interpolation(squeeze(ir(1:3,2,:))',x0(:,1:3),xs);
-    end
+% Leave out impulse responses with weights smaller than prec
+ir = ir(weights>=prec,:,:);
+weights_new = weights(weights>=prec);
+x0_new = x0(weights>=prec,:);
+
+% === IR interpolation ===
+if ~useinterpolation || length(weights)==1
+    ir_new = ir;
+elseif useinterpolation
+    switch interpolationmethod
+    case 'simple'
+        ir_new = sum(bsxfun(@times,ir,weights),1);
+    case 'freqdomain'
+        % See Itoh (1982), Hartung et al. (1999)
+        %
+        % Upsample to avoid phase aliasing in unwrapping of phase
+        TF = fft(ir,4*size(ir,3),3);
+        % Magnitude and phase will be interpolated separately
+        magnitude = abs(TF);
+        phase = unwrap(angle(TF),[],3);
+        % Calculate interpolation only for the first half of the spectrum
+        % and only for original bins
+        idx_half = floor(size(TF,3)/2)+1;
+        magnitude_new = sum(bsxfun(@times,magnitude(:,:,1:4:idx_half),weights),1);
+        phase_new = sum(bsxfun(@times,phase(:,:,1:4:idx_half),weights),1);
+        % Calculate interpolated impulse response from new magnitude and phase
+        ir_new = ifft(magnitude_new.*exp(1i*phase_new),size(ir,3),3,'symmetric');
```

Comment:
Really? This option was already in our code before. The spectrum has to be mirrored manually which means some additional lines of code. Shall I do that?